### PR TITLE
✨ add ads: baidu

### PR DESCRIPTION
**Pull request by PengXing**
Ad an ads: `baidu`.

Baidu is an advertising provider in China, and it plays an important role.

### CODE REVIEW

**Comment by PengXing** [comment]
The limitation of bundlesize of `v0.js` is so strict. I don't know what I can do now. Maybe I should @choumx and @jridgewell.

```
[08:59:33] FAIL  ./dist/v0.js: 83.23KB > maxSize 83.2KB (gzip)
[08:59:33] ERROR: bundlesize found that ./dist/v0.js has exceeded its size cap of 83.2KB (Δ +0.03KB)
[08:59:33] This is part of a new effort to reduce AMP's binary size (#14392).
[08:59:33] Please contact @choumx or @jridgewell for assistance.
```

**Comment by choumx** [comment]
This PR shouldn't affect the size of v0.js. This may be related:

```
[08:59:33] Branch point from master is fa99cee
[08:59:33] WARNING: Failed to retrieve bundle size of fa99cee
[08:59:33] Falling back to comparing to the max bundle size only
```

I restarted the build for you. @danielrozenberg FYI.

**Comment by danielrozenberg** [comment]
Re bundle-size - the base commit on `master` failed to store the new bundle size because, ironically, it increased the bundle size and failed. This isn't how `gulp bundle-size` should behave when executed on a `master` commit (it shouldn't reject the build, it should still store it) but since this is a rare case and I'm already working on changing the behaviour of bundle size checks (#19146,) we'll have to skip this one.

To unblock you we're increasing the max bundle size (https://github.com/ampproject/amphtml-build-artifacts/pull/7), and will manually re-trigger Travis builds on the master branch to create files to store those base sizes for each of the commits that failed so far. Once that's done, we can re-trigger Travis here and it should pass. This should take about ~1-2 hours. Apologies for the hassle!

**Comment by danielrozenberg** [comment]
Update: `bundle-size` now passes, but your PR fails on validator tests. PTAL

File: ads/_config.js [hunk 1]
```diff
@@ -202,6 +202,10 @@ export const adConfig = {
     ],
   },
 
+  'baidu': {
+    prefetch: 'https://dup.baidustatic.com/js/dm.js',
+  },
```
Comment (lannka) [inline ads/_config.js]: could you please implement the render-start API?

full [guideline](https://github.com/ampproject/amphtml/blob/master/ads/README.md#developer-guidelines-for-a-pull-request)

**Comment by PengXing** [comment]
@lannka The `renderStart` and `noContentAvailable` API have been implemented in the latest commit. You may check it.

File: ads/baidu.js [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+  });
+
+  loadScript(
+      global,
+      'https://dup.baidustatic.com/js/dm.js',
+      () => {
+        global.context.renderStart();
+      },
+      () => {
+        global.context.noContentAvailable();
```
Comment (lannka) [inline ads/baidu.js]: this only gets called when the script is failed to load.

does dm.js script fire global events when ad response is back? that will be the time to call renderStart or noContentAvailable, depending on if the ad is filled or not.

**Comment by lannka** [comment]
does the message tell parent if the ad request returns an ad or not? noContentAvailable should be called when the response is empty.

**Comment by lannka** [comment]
Got you, seems tricky then. Would you add a comment to explain that there is no good way to do noContentAvailable call.

Otherwise the PR looks good to me. Thanks!

**Comment by PengXing** [comment]
`noContentAvailable` should be called if parent iframe receives no message. I can use `setTimeout`, but it's not that reliable, due to complicated and unstable network conditions. So, there is no good way to do `noContentAvailable` call perfectly.

**Comment by lannka** [comment]
Thanks @PengXing . I meant to have that as JS comment in the code to inform other developers.

**Comment by PengXing** [comment]
@lannka Please see https://github.com/ampproject/amphtml/pull/19489/commits/da772ceb62ede0969b72fc38003a9a8e5f04036e.

File: 3p/integration.js [hunk 3]
```diff
@@ -250,6 +250,7 @@ import {zedo} from '../ads/zedo';
 import {zen} from '../ads/zen';
 import {zergnet} from '../ads/zergnet';
 import {zucks} from '../ads/zucks';
+import {baidu} from '../ads/baidu';
```
Comment (lannka) [inline 3p/integration.js]: alphabetical order pls

File: 3p/integration.js [hunk 4]
```diff
@@ -479,6 +480,7 @@ register('zedo', zedo);
 register('zen', zen);
 register('zergnet', zergnet);
 register('zucks', zucks);
+register('baidu', baidu);
```
Comment (lannka) [inline 3p/integration.js]: alphabetical order

File: validator/testdata/feature_tests/ads.html [hunk 5]
```diff
@@ -1336,5 +1337,11 @@ <h2>Zucks</h2>
       type="zucks"
       data-frame-id="_bda46cf5ac">
   </amp-ad>
+
+  <h2>Baidu</h2>
```
Comment (lannka) [inline validator/testdata/feature_tests/ads.html]: alphabetical order pls

File: examples/ads.amp.html [hunk 6]
```diff
@@ -235,6 +235,7 @@
         <option>zen</option>
         <option>zergnet</option>
         <option>zucks</option>
+        <option>badiu</option>
```
Comment (lannka) [inline examples/ads.amp.html]: one more here

File: examples/ads.amp.html [hunk 7]
```diff
@@ -1911,5 +1912,11 @@ <h2>Zucks</h2>
       data-frame-id="_bda46cf5ac">
   </amp-ad>
 
+  <h2>Baidu</h2>
```
Comment (lannka) [inline examples/ads.amp.html]: and here

**Comment by PengXing** [comment]
I have completed the modification of all alphabetical order problems. Please check it.

The Travis CI build failed caused by failure of dead link check. And the link of [baidu ad union website](http://union.baidu.com/product/prod-cpro.html) is not dead. Maybe the Travis CI build should be re-run.

```
> [ LinkCheckResult {
    link: 'http://union.baidu.com/product/prod-cpro.html',
    statusCode: 200,
    err: null,
    status: 'alive' } ]
```

**Comment by lannka** [comment]
Tests passed. Merged.

**Comment by PengXing** [comment]
Thanks a lot.

**Pull request merged**